README.md
<!-- Header -->
<!--<h3 align="center">Project Title</h3>-->
<!--<h2 align="center">Project Description</h2>-->

<p align="center">
	<img src ="https://github-readme-stats-jameslafritz.vercel.app/api/pin?username=JamesLaFritz&repo=GraphViewBehaviorTree&theme=react" alt="Git Repo Title and Info" title="Repo Info"/>'
	<br />
	<a href="https://github.com/JamesLaFritz/GraphViewBehaviorTree/issues">Report Bug</a>
        ·
        <a href="https://github.com/JamesLaFritz/GraphViewBehaviorTree/issues">Request Feature</a>
</p>


<!-- PROJECT SHIELDS -->
<p align="center">
  <a href="https://github.com/JamesLafritz/GraphViewBehaviorTree/graphs/contributors">
	  <img src="https://img.shields.io/github/contributors/JamesLafritz/GraphViewBehaviorTree.svg?style=for-the-badge" title="forks Badge" alt="forks Badge"/>
  </a>
  <a href="https://img.shields.io/github/forks/JamesLafritz/GraphViewBehaviorTree.svg?style=for-the-badge">
	  <img src="https://img.shields.io/github/forks/JamesLafritz/GraphViewBehaviorTree.svg?style=for-the-badge" title="stargazers Badge" alt="stargazers Badge"/>
  </a>
  <a href="https://github.com/JamesLafritz/GraphViewBehaviorTree/stargazers">
	  <img src="https://img.shields.io/github/stars/JamesLafritz/GraphViewBehaviorTree.svg?style=for-the-badge" title="Contributors Badge" alt="Contributors Badge"/>
  </a>
  <a href="https://github.com/JamesLafritz/GraphViewBehaviorTree/issues">
	  <img src="https://img.shields.io/github/issues/JamesLafritz/GraphViewBehaviorTree.svg?style=for-the-badge" title="issues Badge" alt="issues Badge"/>
  </a>
  <a href="https://img.shields.io/github/license/JamesLafritz/GraphViewBehaviorTree.svg?style=for-the-badge">
	  <img src="https://img.shields.io/github/license/JamesLafritz/GraphViewBehaviorTree.svg?style=for-the-badge" title="License Badge" alt="License Badge"/>
  </a>
</p>

<!-- Links -->
<p align="center">
  <a href="https://jameslafritz.intensive.gamedevhq.com/">
	  <img src="https://img.shields.io/badge/Portfolio-21759B?style=for-the-badge&logo=wordpress&logoColor=white" title="Portfolio Badge" alt="Portfolio"/>
  </a>
  <a href="https://ktmarine1999.medium.com/">
	  <img src="https://img.shields.io/badge/Articles-000000?style=for-the-badge&logo=medium&logoColor=white" title="medium Badge" alt="medium"/>
  </a>
  <a href="https://www.linkedin.com/in/james-lafritz/">
	  <img src="https://img.shields.io/badge/LinkedIn-0A66C2?style=for-the-badge&logo=linkedin&logoColor=white" title="LinkedIn Badge" alt="LinkedIn"/>
  </a> 
  <a href="https://ktmarine1999.itch.io/">
	  <img src="https://img.shields.io/badge/Itch-fa5c5c.svg?style=for-the-badge&logo=Itch.io&logoColor=white" title="Itch.io Badge" alt="Itch.io"/>
  </a> 
</p>


<!-- PROJECT LOGO -->
<p align="center">
  <a href="https://github.com/JamesLaFritz/GraphViewBehaviorTree">
    <img src="Documentation~/Images/Logo.png" alt="Logo" width="256"/>
  </a>
</p>

<!-- TABLE OF CONTENTS -->
<details open="open">
  <summary><h2 style="display: inline-block">Table of Contents</h2></summary>
  <ol>
    <li>
      <a href="#about-the-project">About The Project</a>
        <ul>
            <li> Node Types </li>
            <ul>
                <li><a href="#decorator-node">Decorator Node</a></li>
            </ul>
            <ul>
                <li><a href="#composite-node">Composite Node</a></li>
            </ul>
	        <ul>
                <li><a href="#action-node">Action Node</a></li>
            </ul>
        </ul>
	    <ul>
            <li><a href="#built-with">Built With</a></li>
        </ul>
        <ul>
            <li><a href="#articles">Articles</a></li>
        </ul>
    </li>
    <li><a href="#license">License</a></li>
    <li><a href="#acknowledgements">Acknowledgements</a></li>
  </ol>
</details>



<!-- ABOUT THE PROJECT -->
## About The Project

![Product Name Screen Shot](Documentation~/Images/ScreenShot.gif)

I had code on sitting around on my hard drive for a Behavior Tree Editor That inspired me to create a bBehavior Tree Editor and write Articles about Behavior Tree and Using the Unity Editor UI Builder.
It has been [Pointed Out](https://github.com/JamesLaFritz/GraphViewBehaviorTree/issues/1) to me very polite and professional that this code is actually from the KIWI Coder.

[![The Kiwi Coder](https://thekiwicoder.com/wp-content/uploads/2020/02/site_logo.png)](https://thekiwicoder.com/)

Behavior Tree using UI Builder, GraphView, and Scriptable Objects in Unity 2021.3

This can be used as a template to build any type on Node Based Editor in Unity. i.e Dialog System or an AI Behavior Tree.
If using for AI Behavior tree I would probably use Single Input Ports not Multiple.
Even though the Nodes themselves are Scriptable Objects this can be changed and there is nothing that states that you have to use Scriptable Objects as the type for the nodes, this was just a decision that I made as it is easier to Bing the properties off the Node. 
Also there is nothing sating that the Nodes have to be stored in the Asset itself as I have done. 

Behavior Tree is an execution tree (uses the [Strategy pattern](https://blog.devgenius.io/strategy-pattern-in-unity-b82065aaa969)) and always starts with a Root Node. This will be so
that the behavior tree knows where to start. Root Node has only one child. Now there are 3 basic main type of nodes
Decorator, Composite Node and Action Node.
Each Node can return one of three states Running, Success, or Failure.
All nodes will be saved in Unity as Scriptable Objects. A Behavior tree will be a Scriptable Object containing all the
Nodes in it.

[Documentation](https://jameslafritz.github.io/GraphViewBehaviorTree/api/GraphViewBehaviorTree.html)

#### Decorator Node
* Has one child and is capable of augmenting the return state of it's child. This uses the [Decorator
pattern](https://blog.devgenius.io/the-decorator-pattern-in-unity-6791ab10b64).

#### Composite Node
* Has a list of children and is the control flow of the behavior tree like switch statements and for
loops. There are 2 types Composite Nodes the Selector and Sequence. This uses the [Composite pattern](https://blog.devgenius.io/composite-pattern-in-unity-fc90e60c946f).

#### Action Node
* The Leaf of the tree, has no children, and is where all of the logic gets implemented.


### Built With

* <a href="https://www.linkedin.com/in/james-lafritz/"><img src="https://img.shields.io/badge/Unity-100000?style=for-the-badge&logo=unity&logoColor=white"/></a>

<!-- Installation -->
### Installation
To install this into your project for modifications
Clone or Download the code from Git Hub into the Asset Folder of your project.
Change the Folder From GraphView Behavior Tree to mach what you wold like.
Edit all Files and change the NameSpace to match your new Folder Name. Do Not forget to change this in the UXML Documents as well. i.e "GraphViewBehaviorTree.Editor.SplitView"
Delete the Runtime/james.lafritz.GraphViewBehaviorTree and Editor/james.lafritz.GraphViewBehaviorTree.Editor Assembly Definition Files, Or Edit/Replace them with your own.

To use Template as is Use the Package manager to Install package from Git Hub. 
see [Creating custom packages for use in Unity](https://blog.devgenius.io/creating-custom-packages-for-use-in-unity-7dfbaa49e4b4)

* In Unity open the package manager
    * Go to Window->PackageManager
    * ![OpenPackageManager](Documentation~/Images/OpenPackageManager.gif)
* Install this Package from git url
    * Click the Plus Drop down
    * Select "add package from git URL"
        * https://github.com/JamesLaFritz/GraphViewBehaviorTree.git
        * Select add
    * ![InstallPackage](Documentation~/Images/InstallPackage.gif)

<!-- Articles -->
### Articles

[Behavior Trees In Unity](https://ktmarine1999.medium.com/behavior-trees-in-unity-20a738b5508c)

[Using Unity’s UI Builder: Basic Set Up](https://blog.devgenius.io/using-unitys-ui-builder-a86faf17bf27)

[Using Unity’s UI Builder: Serialized Object data binding](https://blog.devgenius.io/using-unitys-ui-builder-bc058e1c7d17)

[Using Unity’s UI Builder: Adding Custom Controls](https://blog.devgenius.io/using-unitys-ui-builder-5e793a90a5ae)

[Using Unity’s Graph View: Adding and Deleting Nodes](https://blog.devgenius.io/using-unitys-graph-view-e9fb8e78980e)

[Using Unity’s Graph View: Node Position, Code/Editor Window Reload](https://ktmarine1999.medium.com/using-unitys-graph-view-18b38a23dea5)

[Using Unity’s Graph View: Connecting Nodes in The Editor](https://ktmarine1999.medium.com/4cc55b704548)

[Inspector View in an Editor Window](https://medium.com/@ktmarine1999/inspector-view-in-an-editor-window-5debd86cbe6f)


<!-- ROADMAP -->
## Roadmap

See the [open issues](https://github.com/JamesLaFritz/GraphViewBehaviorTree/issues) for a list of proposed features (and known issues).



<!-- LICENSE -->
## License

Distributed under the MIT License. See `LICENSE` for more information.


<!-- ACKNOWLEDGEMENTS -->
## Acknowledgements

[![The Kiwi Coder Behavior Tree Editor](https://thekiwicoder.com/wp-content/uploads/2021/07/behaviour_tree-2.jpg)](https://thekiwicoder.com/behaviour-tree-editor/)
[![The Kiwi Coder](https://thekiwicoder.com/wp-content/uploads/2020/02/site_logo.png)](https://thekiwicoder.com/)


<!--
Repo Card Exclusive Options:
    show_owner - Show the repo's owner name (boolean)

Common Options:
    title_color - Card's title color (hex color)
    text_color - Body text color (hex color)
    icon_color - Icons color if available (hex color)
    border_color - Card's border color (hex color). (Does not apply when hide_border is enabled)
    bg_color - Card's background color (hex color) or a gradient in the form of angle,start,end
    hide_border - Hides the card's border (boolean)
    theme - name of the theme, choose from all available themes
    cache_seconds - set the cache header manually (min: 1800, max: 86400)
    locale - set the language in the card (e.g. cn, de, es, etc.)
    border_radius - Corner rounding on the card_
Gradient in bg_color

You can provide multiple comma-separated values in bg_color option to render a gradient, the format of the gradient is :-

&bg_color=DEG,COLOR1,COLOR2,COLOR3...COLOR10

Avaliable Repo Card Themes
default_repocard
dark
radical
merko
gruvbox
tokyonight
onedark
cobalt
synthwave
highcontrast
dracula
prussian
monokai
vue
vue-dark
shades-of-purple
nightowl
buefy
blue-green
algolia
great-gatsby
darcula
bear
solarized-dark
solarized-light
chartreuse-dark
nord
gotham
material-palenight
graywhite
vision-friendly-dark
ayu-mirage
midnight-purple
calm
flag-india
omni
react
jolly
maroongold
yeblu
blueberry
slateorange
kacho_ga
outrun
-->


CHANGELOG.md
---- 
* Hash: 277a853 Date: 2022-05-04  Message: [Initial Commit.](https://github.com/JamesLaFritz/GraphViewBehaviorTree/277a8538047101fd1d14af80ddada713b382fc3d) 
----
---- 
* Hash: d3c2bb8 Date: 2022-05-05  Message: [Created Abstract Node Scriptable Object.](https://github.com/JamesLaFritz/GraphViewBehaviorTree/d3c2bb8e7ffaaaad9793234a8675c148a2bd2a9f) 
----
---- 
* Hash: 873dde4 Date: 2022-05-05  Message: [Behavior Tree Created.](https://github.com/JamesLaFritz/GraphViewBehaviorTree/873dde436f3a64906652cdc2d8cf608e9fb657df) 
----
---- 
* Hash: de54b5d Date: 2022-05-05  Message: [Action Node Created.](https://github.com/JamesLaFritz/GraphViewBehaviorTree/de54b5d625c10160a0de848986a2ca448ff60449) 
----
---- 
* Hash: 5174c06 Date: 2022-05-05  Message: [Decorator Node Created.](https://github.com/JamesLaFritz/GraphViewBehaviorTree/5174c06ff608f907bd06a8f2887a96bfdb874090) 
----
---- 
* Hash: 1ae13db Date: 2022-05-13  Message: [Composite Node Added](https://github.com/JamesLaFritz/GraphViewBehaviorTree/1ae13db3a6ecb5a57eff5f1074c3dc4b38442f90) 
----
---- 
* Hash: 425ffe1 Date: 2022-05-13  Message: [Fixed Tree so it does not Continue to run forever.](https://github.com/JamesLaFritz/GraphViewBehaviorTree/425ffe10ab900bab0b0c6e5c78a18b68add46927) 
----
---- 
* Hash: 60f23b5 Date: 2022-05-13  Message: [Corrected the State so the nodes state does not have to be public.](https://github.com/JamesLaFritz/GraphViewBehaviorTree/60f23b5ae1059df7c5f265766d2a554e79742b4c) 
----
---- 
* Hash: 2161e24 Date: 2022-05-13  Message: [First Action Node - Debug Log Node Created](https://github.com/JamesLaFritz/GraphViewBehaviorTree/2161e247ae5121161f4d27308473d8e1a5103ee6) 
----
---- 
* Hash: 7216b30 Date: 2022-05-13  Message: [Behavior Tree Runner Created](https://github.com/JamesLaFritz/GraphViewBehaviorTree/7216b3092315f72886351b6ae0034facf7b29517) 
----
---- 
* Hash: e3f0a6d Date: 2022-05-13  Message: [Created A Folder to store all the Action Nodes](https://github.com/JamesLaFritz/GraphViewBehaviorTree/e3f0a6da1e0fe3b465110146113d00abc38fb9f3) 
----
---- 
* Hash: 0206f4e Date: 2022-05-13  Message: [First Decorator Node Created - Repeat Node - Causes the Child Node to repeat for ever because it returns state running](https://github.com/JamesLaFritz/GraphViewBehaviorTree/0206f4e415284c6bb1e3edbf3cd12b156dd8dd72) 
----
---- 
* Hash: 615be2e Date: 2022-05-13  Message: [Created First Composite Node - Sequence Node](https://github.com/JamesLaFritz/GraphViewBehaviorTree/615be2ed7434540431d3e6632c02accf04ef6ca8) 
----
---- 
* Hash: 7326865 Date: 2022-05-14  Message: [Wait Action Node Added](https://github.com/JamesLaFritz/GraphViewBehaviorTree/7326865547c117718efd2277093e716e75ecc6b3) 
----
---- 
* Hash: 30c92fa Date: 2022-05-15  Message: [Gave Credit to Kiwi Coder](https://github.com/JamesLaFritz/GraphViewBehaviorTree/30c92fa3cd3dfb3a16cf51124f3eaf8880b0bac2) 
----
---- 
* Hash: e0e0059 Date: 2022-05-15  Message: [Created and Set Up an Empty Editor Window to be used as the Behavior Tree Editor](https://github.com/JamesLaFritz/GraphViewBehaviorTree/e0e0059441c23f72f18f996b8d34c9f8b189b18e) 
----
---- 
* Hash: 5cfc7f8 Date: 2022-05-15  Message: [Add the GraphViewBehaviorTree.Editor NameSpace](https://github.com/JamesLaFritz/GraphViewBehaviorTree/5cfc7f830c329f67748f2ab0669ea3c3d1217e04) 
----
---- 
* Hash: 206b7a9 Date: 2022-05-15  Message: [Created Graph View Custom Control that can be used in UI Builder](https://github.com/JamesLaFritz/GraphViewBehaviorTree/206b7a97b1eaab6490829fd581e909890bc14fe6) 
* ----
---- 
* Hash: da4d464 Date: 2022-05-15  Message: [Add Split View Custom Control](https://github.com/JamesLaFritz/GraphViewBehaviorTree/da4d464ebf72c39fba11600f497039247b1f6ec3) 
----
---- 
* Hash: e62598c Date: 2022-05-15  Message: [Used UI Builder to create the Behavior Tree Editor](https://github.com/JamesLaFritz/GraphViewBehaviorTree/e62598c9f916f465079b435b3fb8190627e57a7d) 
----
---- 
* Hash: 973ea80 Date: 2022-05-15  Message: [Changed Nodes to be regular Classes](https://github.com/JamesLaFritz/GraphViewBehaviorTree/973ea803f183fe51b7a67006e02a668aeb8c6bd0) 
----
---- 
* Hash: 8689a84 Date: 2022-05-15  Message: [Added Node View and a way to Display the Node View.](https://github.com/JamesLaFritz/GraphViewBehaviorTree/8689a849f0d773d9fa74b5fcf6bc01635428f07b) 
----
---- 
* Hash: 7320b54 Date: 2022-05-15  Message: [Changed Node To Unity Object so we can bind it to our properties in the editor](https://github.com/JamesLaFritz/GraphViewBehaviorTree/7320b545535e96580227ada27846d9fe0acc1ec4) 
----
---- 
* Hash: c6ecfd6 Date: 2022-05-16  Message: [Text Name Field Added to Tree for Bindable to the Tree Name](https://github.com/JamesLaFritz/GraphViewBehaviorTree/c6ecfd699e4f82c48eb7cb98d790a23145cfef48) 
----
---- 
* Hash: 7af2031 Date: 2022-05-16  Message: [Set to Read Only to Prevent from accidentally changing the name of the Behavior Tree](https://github.com/JamesLaFritz/GraphViewBehaviorTree/7af2031e597df95cae7f6740a8cabbaa589ef1d9) 
----
---- 
* Hash: 26dcb8e Date: 2022-05-24  Message: [Split View Added](https://github.com/JamesLaFritz/GraphViewBehaviorTree/26dcb8e49fd9ff9aea3284f82c2e3c45d383fe83) 
----
---- 
* Hash: 1d36af1 Date: 2022-05-25  Message: [Fixed Node To Be Unity Object](https://github.com/JamesLaFritz/GraphViewBehaviorTree/1d36af1cd06ec3913135fadc8780cf11c86b575f) 
* ----
---- 
* Hash: 6067de6 Date: 2022-05-25  Message: [Fixed Node to be a Unity Object](https://github.com/JamesLaFritz/GraphViewBehaviorTree/6067de6d17bdd525268ec1ed5045b8a59b354ec8) 
* ----
---- 
* Hash: 0ee39f0 Date: 2022-05-28  Message: [Update Node.cs](https://github.com/JamesLaFritz/GraphViewBehaviorTree/0ee39f0ca2ddd3cc48dba5128d5285b14e26912b) 
* ----
---- 
* Hash: 6586bf0 Date: 2022-05-28  Message: [Update WaitNode.cs](https://github.com/JamesLaFritz/GraphViewBehaviorTree/6586bf098644d5e7b9102ac5abaaf0e180277641) 
----
---- 

Editor/BehaviorTreeView.cs
// BehaviorTreeView.cs
// 05-15-2022
// James LaFritz

using System;
using System.Collections.Generic;
using System.Linq;
using UnityEditor;
using UnityEngine.UIElements;
using UnityEditor.Experimental.GraphView;
using UnityEngine;

namespace GraphViewBehaviorTree.Editor
{
    /// <summary>
    /// A View for the Behavior Tree, derived from <see cref="UnityEditor.Experimental.GraphView.GraphView"/>
    /// Can be used in the UI Builder.
    /// </summary>
    public class BehaviorTreeView : GraphView
    {
        public new class UxmlFactory : UxmlFactory<BehaviorTreeView, UxmlTraits> { }

        /// <summary>
        /// Notifies the Observers that a <see cref="Node"/> has been Selected and pass the <see cref="Node"/> that was selected.
        /// </summary>
        public Action<Node> onNodeSelected;

        private BehaviorTree m_tree;
        private bool m_hasTree;

        /// <summary>
        /// Creates a new <see cref="BehaviorTreeView"/>.
        /// </summary>
        public BehaviorTreeView()
        {
            style.flexGrow = 1;
            Insert(0, new GridBackground() { name = "grid_background" });
            this.AddManipulator(new ContentZoomer());
            this.AddManipulator(new ContentDragger());
            this.AddManipulator(new SelectionDragger());
            this.AddManipulator(new RectangleSelector());

            Undo.undoRedoPerformed += UndoRedoPerformed;
        }

        private void UndoRedoPerformed()
        {
            PopulateView(m_tree);
            AssetDatabase.SaveAssets();
        }

        /// <summary>
        /// Populate the View wih the passed in tree
        /// </summary>
        /// <param name="tree">The <see cref="BehaviorTree"/> to populate the View from</param>
        public void PopulateView(BehaviorTree tree)
        {
            m_tree = tree;

            graphViewChanged -= OnGraphViewChanged;
            DeleteElements(graphElements);
            graphViewChanged += OnGraphViewChanged;

            m_hasTree = m_tree != null;
            if (!m_hasTree) return;
            m_tree.GetNodes().ForEach(CreateNodeView);

            foreach (Edge edge in from node in m_tree.GetNodes()
                                  let parentView = GetNodeByGuid(node.guid) as BehaviorTreeNodeView
                                  where parentView is { output: { } }
                                  from child in m_tree.GetChildren(node)
                                  where child != null
                                  let childView = GetNodeByGuid(child.guid) as BehaviorTreeNodeView
                                  where childView is { input: { } }
                                  select parentView.output.ConnectTo(childView.input))
            {
                AddElement(edge);
            }
        }

        /// <summary>
        /// Hook into the Graph View Change to delete Nodes when the Node View Element is slated to be Removed.
        /// </summary>
        /// <param name="graphViewChange"><a href="https://docs.unity3d.com/2021.3/Documentation/ScriptReference/Experimental.GraphView.GraphViewChange.html">GraphViewChange</a></param>
        private GraphViewChange OnGraphViewChanged(GraphViewChange graphViewChange)
        {
            if (graphViewChange.elementsToRemove != null)
            {
                foreach (GraphElement element in graphViewChange.elementsToRemove)
                {
                    switch (element)
                    {
                        case BehaviorTreeNodeView nodeView:
                            DeleteNode(nodeView.node);
                            break;
                        case Edge edge:
                        {
                            BehaviorTreeNodeView parentView = edge.output.node as BehaviorTreeNodeView;
                            BehaviorTreeNodeView childView = edge.input.node as BehaviorTreeNodeView;
                            m_tree.RemoveChild(parentView.node, childView.node);
                            int count = (childView.input.connections ?? Array.Empty<Edge>()).Count();
                            childView.node.hasMultipleParents = count > 2;

                            break;
                        }
                    }
                }
            }

            if (graphViewChange.edgesToCreate != null && m_hasTree)
            {
                foreach (Edge edge in graphViewChange.edgesToCreate)
                {
                    BehaviorTreeNodeView parentView = edge.output.node as BehaviorTreeNodeView;
                    BehaviorTreeNodeView childView = edge.input.node as BehaviorTreeNodeView;

                    m_tree.AddChild(parentView.node, childView.node);
                    parentView?.SortChildren();

                    int count = (childView.input.connections ?? Array.Empty<Edge>()).Count();
                    childView.node.hasMultipleParents = count > 0;
                }
            }

            if (graphViewChange.movedElements != null)
            {
                foreach (BehaviorTreeNodeView parentNodeView
                         in from movedElement in graphViewChange.movedElements
                            let movedNode = movedElement as BehaviorTreeNodeView
                            where movedNode is { input: { connections: { } } }
                            from edge in movedNode.input.connections
                            where edge.output is { node: BehaviorTreeNodeView }
                            select edge.output?.node as BehaviorTreeNodeView)
                {
                    parentNodeView?.SortChildren();
                }
            }

            return graphViewChange;
        }

        /// <summary>
        /// Adds a <see cref="BehaviorTreeNodeView"/> from the passed in Node.
        /// </summary>
        /// <param name="node">The <see cref="Node"/> to create a view for.</param>
        private void CreateNodeView(Node node)
        {
            BehaviorTreeNodeView nodeView = new BehaviorTreeNodeView(node);
            nodeView.onNodeSelected = onNodeSelected;
            if (m_hasTree)
                nodeView.onSetRootNode = _ => m_tree.rootNode = node;
            AddElement(nodeView);
        }

        /// <summary>
        /// Create a new <see cref="Node"/> with a Node View.
        /// </summary>
        /// <param name="type">The Type of Node to create.</param>
        private void CreateNode(Type type)
        {
            if (!m_hasTree) return;

            Node node = m_tree.CreateNode(type);
            CreateNodeView(node);
            Undo.RecordObject(m_tree, "Behavior Tree (Create Node)");

            if (Application.isPlaying) return;

            AssetDatabase.AddObjectToAsset(node, m_tree);
            AssetDatabase.SaveAssets();

            Undo.RegisterCreatedObjectUndo(node, "Behavior Tree (Create Node)");
            EditorUtility.SetDirty(node);
        }

        /// <summary>
        /// Delete a <see cref="Node"/> from the tree.
        /// </summary>
        /// <param name="node">The <see cref="Node"/> to Delete.</param>
        private void DeleteNode(Node node)
        {
            if (!m_hasTree) return;

            m_tree.DeleteNode(node);
            Undo.RecordObject(m_tree, "Behavior Tree (Delete Node)");

            Undo.DestroyObjectImmediate(node);
            AssetDatabase.SaveAssets();
            EditorUtility.SetDirty(m_tree);
        }

        public void UpdateNodeStates()
        {
            foreach (BehaviorTreeNodeView nodeView in nodes.OfType<BehaviorTreeNodeView>())
            {
                nodeView.UpdateState();
            }
        }

        #region Overrides of GraphView

        /// <inheritdoc />
        public override void BuildContextualMenu(ContextualMenuPopulateEvent evt)
        {
            TypeCache.TypeCollection types = TypeCache.GetTypesDerivedFrom<Node>();
            foreach (Type type in types)
            {
                if (type.IsAbstract) continue;
                evt.menu.AppendAction($"{type.BaseType.Name}/{type.Name}",
                                      _ => CreateNode(type));
            }

            base.BuildContextualMenu(evt);
        }

        /// <inheritdoc />
        public override List<Port> GetCompatiblePorts(Port startPort, NodeAdapter nodeAdapter)
        {
            return ports.ToList()!.Where(endPort =>
                                             endPort.direction != startPort.direction &&
                                             endPort.node != startPort.node &&
                                             endPort.portType == startPort.portType).ToList();
        }

        #endregion
    }
}